In [1]:
import pandas as pd
import numpy as np

In [2]:
payments_file = "./installments_payments.csv"
previous_file = "./previous_application.csv"

In [3]:
df_payments = pd.read_csv(payments_file) 
df_previous = pd.read_csv(previous_file)

In [4]:
df_previous.head()

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

In [5]:
df_100042 = df_previous[df_previous['SK_ID_CURR'] == 100042].copy()

In [6]:
df_100042

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  \
71629       1125423      100042     Consumer loans    10602.450   
369486      1127680      100042         Cash loans    39981.060   
381415      1716930      100042         Cash loans    20150.550   
759421      1643964      100042         Cash loans    13431.600   
1011745     1643959      100042         Cash loans    18619.245   
1011746     2535181      100042         Cash loans    47660.850   
1239501     2639682      100042         Cash loans    19148.985   
1410163     2604725      100042         Cash loans    38934.180   
1555449     2137382      100042    Revolving loans     4500.000   

         AMT_APPLICATION  AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  \
71629            97956.0     95431.5            9796.5          97956.0   
369486          900000.0    978408.0               NaN         900000.0   
381415          315000.0    315000.0               NaN         315000.0   
759421          112500.0    119925.0               0.0         112500.0   
1011745         225000.0    247275.0               0.0         225000.0   
1011746         967500.0   1051789.5               NaN         967500.0   
1239501         225000.0    247275.0               0.0         225000.0   
1410163         900000.0    978408.0               NaN         900000.0   
1555449              0.0     90000.0               NaN              NaN   

        WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  ...  \
71629                      TUESDAY                       14  ...   
369486                      MONDAY                       11  ...   
381415                    THURSDAY                       10  ...   
759421                      MONDAY                       14  ...   
1011745                    TUESDAY                       10  ...   
1011746                     MONDAY                        9  ...   
1239501                     FRIDAY                       14  ...   
1410163                   THURSDAY                       13  ...   
1555449                     FRIDAY                       13  ...   

         NAME_SELLER_INDUSTRY  CNT_PAYMENT  NAME_YIELD_GROUP  \
71629    Consumer electronics         10.0        low_normal   
369486                    XNA         36.0        low_normal   
381415                    XNA         24.0            middle   
759421                    XNA         12.0              high   
1011745                   XNA         18.0            middle   
1011746                   XNA         36.0            middle   
1239501                   XNA         18.0            middle   
1410163                   XNA         36.0        low_normal   
1555449                   XNA          0.0               XNA   

                    PRODUCT_COMBINATION  DAYS_FIRST_DRAWING DAYS_FIRST_DUE  \
71629    POS household without interest            365243.0        -2719.0   
369486                 Cash Street: low                 NaN            NaN   
381415              Cash X-Sell: middle            365243.0        -1675.0   
759421                Cash Street: high            365243.0        -2307.0   
1011745             Cash Street: middle            365243.0        -1901.0   
1011746             Cash X-Sell: middle            365243.0         -649.0   
1239501             Cash Street: middle                 NaN            NaN   
1410163                Cash Street: low            365243.0        -1214.0   
1555449                     Card X-Sell             -2547.0        -2513.0   

        DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
71629                     -2449.0        -2449.0          -2444.0   
369486                        NaN            NaN              NaN   
381415                     -985.0        -1345.0          -1338.0   
759421                    -1977.0        -2007.0          -2003.0   
1011745                   -1391.0        -1691.0          -1688.0   
1011746                     401.0         -649.0           -645.0   
1

## Compare Yield to Quoted Rate

We calculate the actual rate quoted at the moment of taking out the loan and compare it to the yield that was eventually earned on the loan.


In [7]:
# We use the previous loan with ID 2604725

df_2604725 = df_payments[df_payments['SK_ID_PREV'] == 2604725].copy()

In [8]:
df_2604725

SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  \
100789      2604725      100042                     1.0   
469627      2604725      100042                     1.0   
520877      2604725      100042                     1.0   
553719      2604725      100042                     1.0   
631151      2604725      100042                     1.0   
882452      2604725      100042                     2.0   
965673      2604725      100042                     1.0   
978530      2604725      100042                     1.0   
1639695     2604725      100042                     1.0   
1765468     2604725      100042                     1.0   
1863804     2604725      100042                     1.0   
2406607     2604725      100042                     1.0   
2477304     2604725      100042                     1.0   
2590718     2604725      100042                     1.0   
2665523     2604725      100042                     1.0   
2775269     2604725      100042                     1.0   
2840333     2604725      100042                     1.0   
3760605     2604725      100042                     1.0   

         NUM_INSTALMENT_NUMBER  DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  \
100789                       5          -1094.0             -1126.0   
469627                      16           -764.0              -763.0   
520877                      11           -914.0              -945.0   
553719                       6          -1064.0             -1064.0   
631151                       2          -1184.0             -1185.0   
882452                      17           -734.0              -740.0   
965673                       3          -1154.0             -1155.0   
978530                      14           -824.0              -854.0   
1639695                     15           -794.0              -822.0   
1765468                      5          -1094.0             -1092.0   
1863804                      4          -1124.0             -1126.0   
2406607                     13           -854.0              -885.0   
2477304                      9           -974.0              -977.0   
2590718                      7          -1034.0             -1034.0   
2665523                      8          -1004.0             -1005.0   
2775269                      1          -1214.0             -1218.0   
2840333                     12           -884.0              -914.0   
3760605                     10           -944.0              -963.0   

         AMT_INSTALMENT  AMT_PAYMENT  
100789         38934.18      6263.28  
469627         38934.18     38934.18  
520877         38934.18     38934.18  
553719         38934.18     38934.18  
631151         38934.18     38934.18  
882452        642481.83    642481.83  
965673         38934.18     38934.18  
978530         38934.18     38934.18  
1639695        38934.18     38934.18  
1765468        38934.18     32670.90  
1863804        38934.18     38934.18  
2406607        38934.18     38934.18  
2477304        38934.18     38934.18  
2590718        38934.18     38934.18  
2665523        38934.18     38934.18  
2775269        38934.18     38934.18  
2840333        38934.18     38934.18  
3760605        38934.18     38934.18

In [9]:
# What was the quoted rate on this loan?

details_2604725 = df_previous[df_previous['SK_ID_PREV'] == 2604725]
annuity = details_2604725['AMT_ANNUITY'].iloc[0]
credit = details_2604725['AMT_CREDIT'].iloc[0]
no_of_payments = details_2604725['CNT_PAYMENT'].iloc[0]
print("Annuity: %.2f" % annuity)
print("Credit: %.2f" % credit)
print("Number of monthly instalments: %i" % no_of_payments)
PMT = [-credit]
PMT = PMT + ([annuity] * int(no_of_payments))
irr = np.irr(PMT)
ear = (1+irr)**12 - 1
print("Annualized Rate on the loan: %.2f%%" % (100*ear))



Annuity: 38934.18
Credit: 978408.00
Number of monthly instalments: 36
Annualized Rate on the loan: 28.15%


In [11]:
start_date = int(df_2604725['DAYS_INSTALMENT'].min() - 30)
print("Starting day: %i" % start_date)
end_date = int(df_2604725['DAYS_ENTRY_PAYMENT'].max())
print("Last payment on: %i" % end_date)
period_length = end_date - start_date
cf = [0] * (period_length + 1)
cf[0] = -credit
for i, row in df_2604725.iterrows():
    index = int(row['DAYS_ENTRY_PAYMENT'] - start_date)
    amount = row['AMT_INSTALMENT']
    cf[index] = amount
print(cf)
irr = np.irr(cf)
ear = (1+irr)**360 - 1 # we take 1 year to be 360 days
print("Annualized Rate on the loan: %.2f%%" % (100*ear))

Starting day: -1244
Last payment on: -740
[-978408.0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38934.18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

So from above calculations we conclude that for loan with ID 2604725 the rate at which the loan was closed was 28.15% and in the end the yield was 28.72%